### DISCUSSION TITLE
Compile issue with C++19

### DISCUSSION THREAD

**Discussion by maallyn** [jamulussoftware/jamulus#1098] [General]
Are any of you aware that if I use the latest AT Creator and the latest Virtual Studio from Microsoft, which uses C++19, that I am unable to do builds of the current tip of Jamulus?

The following is a posing I put on the Discord Tech Team group, but I was suggested to post it here as well as some of the developers may not be on Discord nor Facebook.

================

I think that we may be having an issue with compiling for Windows using the latest QT Creator
and Microsoft Visual Studio.

It turns out that Microsoft has added a data type starting in C++17, which is std::byte, which
is breaking our compiles where we are using the type byte and it claims that it's ambiguous.

This is discussed in these pages:

https://studiofreya.com/2018/01/06/visual-studio-2017-with-cpp17-and-boost/#stdbyte-ambiguous-symbol-and-rpcndr.h

https://stackoverflow.com/questions/60793009/error-using-qt-during-build-c2872-byte-ambiguous-symbol

https://developercommunity.visualstudio.com/content/problem/93889/error-c2872-byte-ambiguous-symbol.html

Looking at other sites on line it appears to me that the QT Community and the Microsoft Visual
Community are in dis-agreement on what to do about this.

The work arounds that I am seeing, based on the comments in the three links I mention here are
to prepend byte with std::byte or to use a define

#define _HAS_STD_BYTE 0 in whichever c programs that we have that use the byte data type.

I know that we are currently using the git method of doing our compiles upon a release. However,
I fear that GIT is out of our control and they may depreciate any C++15 and earlier, which would
force us into C++17.

I am willing to try to slog my way through the code to try to fix this so that it will compile
with the latest tools, which should render us safe with using the git's compile engines if they
do cut off use of the older MSVS C compilers.

And by the way, I never got any responses on Facebook about others doing Windows compiling,
which suggests to me that the only community is here.

**Comment by ann0see** [comment]
Interesting observation. I compile Jamulus on Windows too (msvc2019 build tools) but use the deploy_windows.ps1 script to build the installer too. I must say, that I don’t write much code on Windows and didn’t come across this issue yet. 

Can you have a look at the issues tab (there’s an issue about compilation on Windows)?

**Reply by atsampson** [reply]
The code shouldn't be doing `using namespace std;` in the first place. I'll put in a pull request...

**Comment by maallyn** [comment]
Folks:

I just tried to compile Jamulus Release 3.6.3 with QT 6.0.1 (which is the latest) with msvc2019_64 and am getting some errors.

Without knowing what I should save to here, I grabbed the entire compile output from QT.

`please post logs to pastebin or similar`

**Reply by ann0see** [reply]
Edit. Ah no. Sorry for the ping Peter. 

I assume it’s related to Qt6. One day we‘ll need to move on.

**Reply by atsampson** [reply]
That's actually less broken that I'd expect given it's an entirely new Qt version!

```
cwavestream.h(33): error C2079: 'recorder::secondsAt48K' uses undefined class 'QString'
```

The most likely cause of that kind of error is that `cwavestream.h` uses the `QString` class, but is missing `#include <QString>`. In previous versions of Qt it would have picked it up via a transitive `#include` in another header file it was using. You could try adding the `#include`s it's complaining about, and seeing if it gets further... but be aware that there probably will be other stuff that needs porting for Qt 6.

**Reply by pljones** [reply]
We don't support Qt6.  Qt must be no newer than 5.12.2.  But yes, if that's all that's broken... :)

**Comment by softins** [comment]
Mark, I've just merged atsampson's PR to remove the global import of namespace std.

Please could you fetch the latest upstream master and try compiling again? Please make sure you are using Qt5, not Qt6.

**Comment by maallyn** [comment]
I just did attempt on tip as of about 12 Noon Pacific U.S. Time Thursday Feb 25.

It complains of these four issues:

========================================================================
C:\Users\allyn\Desktop\jamulus-client-working\jamulus-tip\jamulus\windows\ASIOSDK2\host\pc\asiolist.cpp:24: warning: C4267: 'argument': conversion from 'size_t' to 'DWORD', possible loss of data
========================================================================
C:\Users\allyn\Desktop\jamulus-client-working\jamulus-tip\jamulus\windows\ASIOSDK2\host\pc\asiolist.cpp:31: warning: C4267: 'argument': conversion from 'size_t' to 'DWORD', possible loss of data
========================================================================
C:\Users\allyn\Desktop\jamulus-client-working\jamulus-tip\jamulus\libs\opus\celt\modes.c:376: warning: C4244: '=': conversion from 'double' to 'opus_val16', possible loss of data
========================================================================
:-1: warning: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\bin\HostX64\x64\cl.exe" is used by qmake, but "C:\Qt\Tools\QtCreator\bin\clang\bin\clang.exe" is configured in the kit.
Please update your kit (Desktop Qt 5.15.2 MSVC2019 64bit) or choose a mkspec for qmake that matches your target environment better.
========================================================================

I am making a guess that the first two are issues with ASIOSDK2, which is a Microsoft product. I downloaded/installed that only last week.

The third issue appears to be part of our code, someting to do with the opus_val16. A second thought, this may not be us, but the Opus people???

The third one may be an issue with our mkspec? I downloaded and installed QT today; about 1 hour prior to this build attempt.

Here is the output of the compile. 

I am a bit confused, If you look at the very end, it says that is completed okay, so I don't know if the issues that were listed were only warnings or not. 

I am also confused as to where it puts the final executable. I can press run and I see it running, but I don't see the executable in the same place that I would have seen it in a Linux build. 

`please post logs to pastebin or similar`
Edit by @ann0see: See the build output:
[BuildOutput.txt](https://github.com/jamulussoftware/jamulus/files/6051023/BuildOutput.txt)

**Reply by ghost** [reply]
Don't paste junk like this a forum.
Attach the file sepatately.
I cannot find your comments, if you made any.

Now to the guys that manage this forum:
you shoud limit the length of lines and limit the number of lines
to 130 by 130 to force people to add attchments.

**Reply by ann0see** [reply]
Hi David, Hi maalyn,

Yes, this post is rather long but I saw the warnings in bold above the output?

Not sure if it's even possible to give a maximum length or attachment? Also I'm not sure who is the real moderator here? Who did the moderation on SF?

**Reply by gilgongo** [reply]
I've been modding them.

**Reply by maallyn** [reply]
I'm sorry about doing the postings of the logs. I thought that some of the other developers wanted to see the detail. Is the attach files by dragging & dropping, selecting or pasting them the right thing to do or do I post them on my own website and put a link here?

**Reply by ann0see** [reply]
That's not a big problem ;-). I think you can drag and drop files too. Pastebin or GitHub Gist should also work.
I've now uploaded it as a txt file.
